hamster_gtk/misc/widgets/raw_fact_entry.py

```diff
@@ -0,0 +1,229 @@
+        if remove_prefix:
+            result = remove_prefix(self.current_segment, result)
+        return result
+
+
+    # Callbacks
+    def _on_facts_changed(self, evtl):
+        """Callback triggered when facts have changed."""
+        self.set_completion(RawFactCompletion(self._controller))
+
+    def _on_changed(self, widget):
+        """
+        Callback triggered whenever entry text is changed.
+
+        Its main task is to keep track of which segment of the raw fact string the
+        user is currently editing. For this the whole string is inspecred and matched
+        against our regex. By coparing the current cursor position with individual
+        matched segment spans a guess about which one is currently edited is made.
+        """
+        def get_segment(match):
+            """
+            Return the segment the cursor is currently in.
+
+            Returns:
+                str or None: Segment identifier or None if cursor not within any segment.
+            """
+            result = None
+            cursor_position = self.get_position()
+            for segment in ('timeinfo', 'activity', 'category', 'tags', 'description'):
+                start, end = match.span(segment)
+                if start <= cursor_position <= end:
+                    result = segment
+            return result
+
+        def run_autocomplete(segment):
+            completion = self.get_completion()
+            completion.set_model(completion.segment_models[segment])
+
+        self.match = helpers.decompose_raw_fact_string(_u(self.get_text()), raw=True)
+        # Do nothing if we could not 'match' the raw string.
+        # Please note that the completions 'match function' will be run
+        # regardless of what we do here.
+        if self.match:
+            self.current_segment = get_segment(self.match)
+            if self.current_segment in ('activity', 'category'):
+                run_autocomplete(self.current_segment)
+
+
+class RawFactCompletion(Gtk.EntryCompletion):
+    """
+    Return a completion instance to match 'activity@category' strings.
+
+    Returns:
+        Gtk.EntryCompletion: Completion instance.
+    """
+
+    def __init__(self, controller, *args, **kwargs):
+        """instantiate class."""
+        super(RawFactCompletion, self).__init__(*args, **kwargs)
+        self._controller = controller
+        self._activities_model, self._categories_model = self._get_stores()
+        self.set_model(self._activities_model)
+        self.set_text_column(0)
+        self.set_match_func(self._match_anywhere, None)
+        self.connect('match-selected', self._on_match_selected)
+        self.segment_models = {
+            'activity': self._activities_model,
+            'category': self._categories_model,
+        }
+
+    def _get_stores(self):
+        # [FIXME]
+        # This needs to be improved. Right now activity and categories can be
+        # duplicates.
+        activities, categories = [], []
```

Comment:
The original reason was to keep the "order information" so one could use it to prioritize matches. But if we ever implement a more clever matching solution (may in line with legacy hamster) we can just remove the ``set``. So you are right, as it is now its neither here nor there...